streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -785,6 +787,22 @@ private void addTasksToStateUpdater() {
         }
     }
 
+    private void pauseTasksInStateUpdater() {
+        for (final Task task : stateUpdater.getUpdatingTasks()) {
```

Comment:
I'm a bit concerned about perf here since this is called for each iteration, and it will construct a list of read-only tasks hence generating a lot of young gen for GCs, but other than the current approach I have not got to a better solution.

Reply:
Alternative idea is to add a `pauseTasks(topologyMetadata)` / `resumeTasks(topologyMetadata)` to replace the `pauseTask/resumeTask` by task ids, but that would a bit intrusive as introducing topologyMetadata to the state updater..

Reply:
Actually, read-only tasks are currently only generated in `getTasks()` which is not called by `getUpdatingTasks()`. However, if we decide to expose `getUpdatingTasks()` we should also generate read-only tasks there.   

We could pause topologies instead of tasks in the state updater. Instead of `pause(task)` and `resume(task)`, we would then have `pause(topology)` and `resume(topology)` in the state updater interface. Within the default updater we could store the paused topologies and process tasks according to whether they are part of a paused topology or not.

WDYT?

Reply:
Thanks @cadonna that aligns with my alternative idea. I was concerned if exposing the topology metadata inside state updater would be too intrusive, but maybe that's a better trade-off for perf.

I will follow that route in this PR and complete test coverage then.

Reply:
Haha, when I wrote my reply I did not see your alternative idea because the page was not refreshed. That means, we came to similar ideas which is a good sign. 🙂